File: pr2 full/4.cpp
#include <iostream>
#include <string>

#include "func.h"

using namespace std;

void runTask4() {
    int p, g, a, b;
    string message;

    cout << "\n--- Задание 4 ---" << endl;
    cout << "Диффи-Хеллман: получение общего ключа и шифрование текста" << endl;
    cout << "Введите простое число p: ";
    cin >> p;

    if (!isPrime(p)) {
        cout << "Ошибка: p должно быть простым." << endl;
        waitEnter();
        return;
    }

    cout << "Введите основание g (1 < g < p): ";
    cin >> g;
    if (g <= 1 || g >= p) {
        cout << "Ошибка: g должно быть больше 1 и меньше p." << endl;
        waitEnter();
        return;
    }

    cout << "Введите секрет Алисы a: ";
    cin >> a;
    cout << "Введите секрет Боба b: ";
    cin >> b;

    if (a <= 0 || b <= 0) {
        cout << "Ошибка: секретные числа должны быть положительными." << endl;
        waitEnter();
        return;
    }

    cin.ignore();
    cout << "Введите исходный текст: ";
    getline(cin, message);

    int A = powerMod(g, a, p);
    int B = powerMod(g, b, p);
    int keyAlice = powerMod(B, a, p);
    int keyBob = powerMod(A, b, p);

    cout << "Открытое значение Алисы A = g^a mod p = " << A << endl;
    cout << "Открытое значение Боба B = g^b mod p = " << B << endl;
    cout << "Ключ Алисы: " << keyAlice << endl;
    cout << "Ключ Боба: " << keyBob << endl;

    if (keyAlice != keyBob) {
        cout << "Ошибка: ключи не совпали." << endl;
        waitEnter();
        return;
    }

    string encrypted = encryptText(message, keyAlice);
    string decrypted = decryptText(encrypted, keyBob);

    cout << "Зашифрованный текст: " << encrypted << endl;
    cout << "Расшифрованный текст: " << decrypted << endl;

    waitEnter();
}

File: pr2 full/Задание 2 Практика Go/4.go
package main

import (
	"fmt"
	"strings"
)

func runTask4() {
	var p, g, a, b int

	fmt.Println("\n--- Задание 4 ---")
	fmt.Println("Диффи-Хеллман: получение общего ключа и шифрование текста")
	fmt.Print("Введите простое число p: ")
	fmt.Fscan(reader, &p)

	if !isPrime(p) {
		reader.ReadString('\n')
		fmt.Println("Ошибка: p должно быть простым.")
		waitEnter()
		return
	}

	fmt.Print("Введите основание g (1 < g < p): ")
	fmt.Fscan(reader, &g)
	if g <= 1 || g >= p {
		reader.ReadString('\n')
		fmt.Println("Ошибка: g должно быть больше 1 и меньше p.")
		waitEnter()
		return
	}

	fmt.Print("Введите секрет Алисы a: ")
	fmt.Fscan(reader, &a)
	fmt.Print("Введите секрет Боба b: ")
	fmt.Fscan(reader, &b)
	reader.ReadString('\n')

	if a <= 0 || b <= 0 {
		fmt.Println("Ошибка: секретные числа должны быть положительными.")
		waitEnter()
		return
	}

	fmt.Print("Введите исходный текст: ")
	message, _ := reader.ReadString('\n')
	message = strings.TrimRight(message, "\r\n")

	A := powerMod(g, a, p)
	B := powerMod(g, b, p)
	keyAlice := powerMod(B, a, p)
	keyBob := powerMod(A, b, p)

	fmt.Println("Открытое значение Алисы A = g^a mod p =", A)
	fmt.Println("Открытое значение Боба B = g^b mod p =", B)
	fmt.Println("Ключ Алисы:", keyAlice)
	fmt.Println("Ключ Боба:", keyBob)

	if keyAlice != keyBob {
		fmt.Println("Ошибка: ключи не совпали.")
		waitEnter()
		return
	}

	encrypted := encryptText(message, keyAlice)
	decrypted := decryptText(encrypted, keyBob)

	fmt.Println("Зашифрованный текст:", encrypted)
	fmt.Println("Расшифрованный текст:", decrypted)

	waitEnter()
}
